sslocal: link insights ingester into the memory monitor

jasonlmfong opened the pull request:
The sql stats ingester is not currently plugged into the memory monitor, and this has caused
issues in the past with long running transactions having their statements stats occupy memory
while being unaccounted for.

This change introduces the memory monitor into sql stats ingester and uses it to account for
the memory usage for statements.

Resolves: #158453
Epic: None
Release Note: None

dhartunian reviewed (changes requested):
> @dhartunian reviewed 1 file and made 2 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159841)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jasonlmfong).

___
*[`pkg/sql/sqlstats/sslocal/sql_stats_ingester_test.go` line 456 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159841#-Oh68DSP3e2NzR95NAyL:-Oh68DSP3e2NzR95NAyM:bdauhhy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc3c51324009dd1307d54ecfe9f8b4084496dd1/pkg/sql/sqlstats/sslocal/sql_stats_ingester_test.go#L456)):*
> ```Go
> // TestSQLStatsIngesterMemoryAccounting verifies that memory is properly tracked
> // when buffering and flushing statements.
> func TestSQLStatsIngesterMemoryAccounting(t *testing.T) {
> ```

Avoid writing many tests with individual boilerplate like the ones you're adding in this file. It's tough to read and modify and the intent of the test ends up hidden in the large test.

What I like to see is a table-driven test, where you do set up at the top, construct a list of test case structs and execute `t.Run` on them in a loop. Then the actual verification/assertion code is in the loop body and easily segregated visually from the setup.

Then it's obvious that all the cases use the same setup etc. Right now it's very difficult to see how the test cases differ.
___
*[`pkg/sql/sqlstats/sslocal/sql_stats_ingestor.go` line 425 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159841#-Oh68s6w6JIg3f7LJ49q:-Oh68s6w6JIg3f7LJ49r:b-yldqqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc3c51324009dd1307d54ecfe9f8b4084496dd1/pkg/sql/sqlstats/sslocal/sql_stats_ingestor.go#L425)):*
> ```Go
> 
> 		// Release memory for this session's buffered statements.
> 		if i.acc != nil {
> ```

is there value to requiring `i.acc` to be non-nil in the constructor? you can always pass an empy pool or something to all the test cases that are currently nil.


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
> @jasonlmfong made 2 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159841)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian).

___
*[`pkg/sql/sqlstats/sslocal/sql_stats_ingestor.go` line 425 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159841#-Oh68s6w6JIg3f7LJ49q:-Oh6qKL72Cz2-0O-G77E:b1h5ipk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc3c51324009dd1307d54ecfe9f8b4084496dd1/pkg/sql/sqlstats/sslocal/sql_stats_ingestor.go#L425)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

is there value to requiring `i.acc` to be non-nil in the constructor? you can always pass an empy pool or something to all the test cases that are currently nil.
</blockquote></details>

a bit confused by your comment, this is not the constructor though
___
*[`pkg/sql/sqlstats/sslocal/sql_stats_ingester_test.go` line 456 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159841#-Oh68DSP3e2NzR95NAyL:-Oh6qoR0AHIJkUPfU2iR:bbadstf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc3c51324009dd1307d54ecfe9f8b4084496dd1/pkg/sql/sqlstats/sslocal/sql_stats_ingester_test.go#L456)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Avoid writing many tests with individual boilerplate like the ones you're adding in this file. It's tough to read and modify and the intent of the test ends up hidden in the large test.

What I like to see is a table-driven test, where you do set up at the top, construct a list of test case structs and execute `t.Run` on them in a loop. Then the actual verification/assertion code is in the loop body and easily segregated visually from the setup.

Then it's obvious that all the cases use the same setup etc. Right now it's very difficult to see how the test cases differ.
</blockquote></details>

ok, will do that


<!-- Sent from Reviewable.io -->

yuzefovich commented on pkg/sql/sqlstats/sslocal/sql_stats_ingestor.go:
```diff
@@ -393,19 +416,40 @@ func NewSQLStatsIngester(
 
+		// Release memory for this session's buffered statements.
+		if i.acc != nil {
+			if size, ok := i.stmtSizes[sessionID]; ok {
```
Is it ever possible that we had a session in `statementsBySessionID` but don't have the size entry in `stmtSizes`? I would think not.

The same question goes for other spots. What I'm concerned about is whether we can have memory accounting drift somehow, i.e. we'll keep making new allocations with `Grow` calls but they aren't accompanied by the respective `Shrink` calls, and in the end we have accounting leak.

jasonlmfong commented on pkg/sql/sqlstats/ssprovider.go:
```diff
@@ -101,6 +102,41 @@ type RecordedStmtStats struct {
 
+// Size returns the approximate memory footprint of RecordedStmtStats.
+func (r *RecordedStmtStats) Size() int64 {
+	size := int64(unsafe.Sizeof(*r))
```
TIL, thanks
<!-- Reviewable comment -OhAeVvA66BXLGdEu9IU:b-y7y73l -->
<!-- Sent from Reviewable.io -->

jasonlmfong commented on pkg/sql/sqlstats/ssprovider.go:
```diff
@@ -101,6 +102,41 @@ type RecordedStmtStats struct {
+	size += int64(len(r.PlanGist))
+
+	// Account for slices.
+	size += int64(cap(r.Nodes)) * int64(unsafe.Sizeof(int64(0)))
```
ah ok
<!-- Reviewable comment -OhAe_fa-Nli0quxs1OV:b-7tegs0 -->
<!-- Sent from Reviewable.io -->

jasonlmfong commented on pkg/sql/sqlstats/sslocal/sql_stats_ingestor.go:
```diff
@@ -356,12 +374,17 @@ func NewSQLStatsIngester(
 	}
 
+	if parentMon != nil {
+		i.acc = parentMon.MakeConcurrentBoundAccount()
```
I see
<!-- Reviewable comment -OhAfCg8-ESAm5x6H5BW:b-87tsrt -->
<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
> @jasonlmfong made 5 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159841)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @yuzefovich).



<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
> @jasonlmfong made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159841)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @yuzefovich).

___
*[`pkg/sql/sqlstats/sslocal/sql_stats_ingestor.go` line 425 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159841#-Oh68s6w6JIg3f7LJ49q:-OhAqeH149PfqahdFmU2:bx05rde) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc3c51324009dd1307d54ecfe9f8b4084496dd1/pkg/sql/sqlstats/sslocal/sql_stats_ingestor.go#L425)):*
<details><summary><i>Previously, jasonlmfong (Jason Fong) wrote…</i></summary><blockquote>

a bit confused by your comment, this is not the constructor though
</blockquote></details>

I think Yahor's comment above addresses this, I'll add the fallback bound account 


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> @dhartunian reviewed 2 files and all commit messages, made 2 comments, and resolved 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159841)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jasonlmfong and @yuzefovich).

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159841#-OhzaSrV13Qk5g010Tdk:-OhzaSrV13Qk5g010Tdl:bxs5bac):*
nit: wrap
___
*[`pkg/sql/sqlstats/sslocal/sql_stats_ingester_test.go` line 469 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159841#-OhzbLf93aBB6wyPKIqS:-OhzbLf93aBB6wyPKIqT:b46dsrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a97da0a09a80517e1fe832f0c5e5eec6de74279/pkg/sql/sqlstats/sslocal/sql_stats_ingester_test.go#L469)):*
> ```Go
> 			name:        "basic buffering and flushing",
> 			memoryLimit: 10 * 1024, // 10KB
> 			record: func(t *testing.T, ingester *SQLStatsIngester, testSink *sqlStatsTestSink) {
> ```

what you've done with the tests is a good start, but I think having gigantic inline functions in the test cases is also something to avoid.

I think you should try to increase the abstraction level in the test cases and remove the need for inline functions. Try and make the test case struct self-documenting instead of "flexible" like you've done here which obscures intent and requires me to read each case implementation.

The general test structure I'm observing is:

- We need to record some set of pre-determined statements. These should be constructed as `RecordedStmtStats` literals in the test case. Honestly, writing an inline slice of 5 rows is perfectly fine. Or you can pre-construct them above the test cases definition and then use them.
- You should include along with each statement, some expectation of whether it should be recorded into the memory accounting system, so perhaps each statement struct should be a pair of `RecordedStmtStats` and `expectRecorded bool` which tells the test runner whether this statement should succeed in recording or not.
- As we record statements, each successful recording should have its sessionID stored so that we can call `FlushBuffer` on it afterwards and observe the account reduce in size.


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
> @jasonlmfong made 2 comments and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159841)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @yuzefovich).

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159841#-OhzaSrV13Qk5g010Tdk:-Oi-OmV8AosOk7w7W9lb:bb7f2mf):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: wrap
</blockquote></details>

done
___
*[`pkg/sql/sqlstats/sslocal/sql_stats_ingester_test.go` line 469 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159841#-OhzbLf93aBB6wyPKIqS:-Oi-OtP71KAH_1X7BHDM:b1hlobt) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a97da0a09a80517e1fe832f0c5e5eec6de74279/pkg/sql/sqlstats/sslocal/sql_stats_ingester_test.go#L469)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

what you've done with the tests is a good start, but I think having gigantic inline functions in the test cases is also something to avoid.

I think you should try to increase the abstraction level in the test cases and remove the need for inline functions. Try and make the test case struct self-documenting instead of "flexible" like you've done here which obscures intent and requires me to read each case implementation.

The general test structure I'm observing is:

- We need to record some set of pre-determined statements. These should be constructed as `RecordedStmtStats` literals in the test case. Honestly, writing an inline slice of 5 rows is perfectly fine. Or you can pre-construct them above the test cases definition and then use them.
- You should include along with each statement, some expectation of whether it should be recorded into the memory accounting system, so perhaps each statement struct should be a pair of `RecordedStmtStats` and `expectRecorded bool` which tells the test runner whether this statement should succeed in recording or not.
- As we record statements, each successful recording should have its sessionID stored so that we can call `FlushBuffer` on it afterwards and observe the account reduce in size.
</blockquote></details>

ack, made the changes to tests to have a more unified main loop. also added a 4th test case


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 
Just the nit on commit message.

> @dhartunian reviewed all commit messages and made 3 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159841#-:-OiJvgZRCBdiq0eZ3mqT:b-85hjnm)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jasonlmfong and @yuzefovich).

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159841#-OhzaSrV13Qk5g010Tdk:-OiJu4u-BOVfNySwLMPR:b-bygs7f):*
<details><summary><i>Previously, jasonlmfong (Jason Fong) wrote…</i></summary><blockquote>

done
</blockquote></details>

Sorry, I should have been more specific:
- This wrap is too wide. Wrap at <=100 chars.
- The message should be wrapped consistently. The paragraph below is longer but is not wrapped.

Git commit descriptions should be wrapped to 100 characters at maximum width. This is a more natural paragraph width when reading and works nicely in the terminal when you have a multi-column setup and in `git log`.

https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages#General-best-practices%3A
___
*[`pkg/sql/sqlstats/sslocal/sql_stats_ingester_test.go` line 469 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159841#-OhzbLf93aBB6wyPKIqS:-OiJvat40H1UFDtHVL4u:b-oxnsac) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a97da0a09a80517e1fe832f0c5e5eec6de74279/pkg/sql/sqlstats/sslocal/sql_stats_ingester_test.go#L469)):*
<details><summary><i>Previously, jasonlmfong (Jason Fong) wrote…</i></summary><blockquote>

ack, made the changes to tests to have a more unified main loop. also added a 4th test case
</blockquote></details>

Nice, thanks for making this change!


<!-- Sent from Reviewable.io -->

jasonlmfong committed:
sslocal: link insights ingester into the memory monitor

The sql stats ingester is not currently plugged into the memory monitor, and this has caused
issues in the past with long running transactions having their statements stats occupy memory
while being unaccounted for.

This change introduces the memory monitor into sql stats ingester and uses it to account for
the memory usage for statements.

Resolves: #158453
Epic: None
Release Note: None
